File: competition/Mar 7. 2021/Solution5699.java
import java.util.*;

class Solution5699 {
	int[] dist;
	int size;
	Map<Integer, List<int[]>> map;
    long[] mem;
    final int mod = 1000000000 + 7;

    public int countRestrictedPaths(int n, int[][] edges) {
    	size = n;
    	dist = new int[n + 1];
        mem = new long[n + 1];
        Arrays.fill(mem, -1);
    	Arrays.fill(dist, Integer.MAX_VALUE);
    	map = new HashMap<>();
    	for(int i = 1; i <= n; i++){
    		map.put(i, new LinkedList<>());
    	}
    	for(int[] edge: edges){
    		map.get(edge[0]).add(new int[]{edge[1], edge[2]});
    		map.get(edge[1]).add(new int[]{edge[0], edge[2]});
    	}
    	findshortest(n);
        int res = (int) findpaths(1, n);
        return res;
    }

    private long findpaths(int start, int end){
        if(start == end){
            return 1;
        }
        if(mem[start] != -1){
            return mem[start];
        }
        int count = 0;
        for(int[] ele: map.get(start)){
            int point = ele[0];
            if(dist[point] < dist[start]){
                count += findpaths(point, end);
                count = count % mod;
            }
        }
        mem[start] = count;
        return count;
    }

    private void findshortest(int start){
    	boolean[] visit = new boolean[size + 1];
    	dist[start] = 0;
		Comparator<Integer> cmp = (x, y) -> dist[x] - dist[y];
		PriorityQueue<Integer> pq = new PriorityQueue<>(cmp);
    	pq.offer(start);
        while(!pq.isEmpty()){
            int cur = pq.poll();
            if(visit[cur]){
                continue;
            }
            visit[cur] = true;
            for(int[] ele: map.get(cur)){
                int point = ele[0];
                dist[point] = Math.min(dist[point], dist[cur] + ele[1]);
                pq.add(point);
            }
        }
    }

    public static void main(String[] args){
    	Solution5699 s = new Solution5699();
    	int[][] edges = {{1,2,3},{1,3,3},{2,3,1},{1,4,2},{5,2,2},{3,5,1},{5,4,10}};
        s.countRestrictedPaths(5, edges);
        int[] dis = s.dist;
        for(int i = 1; i <= 5; i++){
            System.out.println(dis[i]);
        }
	}
}

File: competition/Mar 7. 2021/Solution5699.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<queue>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;

const int mod = 1000000000 + 7;

struct cmp{
    bool operator()(pii& a, pii& b) const {
        return a.second > b.second;
    }
};

class Solution5699 {
public:
    int size;
    vector<int> dist;
    vector<unordered_map<int, int>> map;
    vector<int> mem;

    int countRestrictedPaths(int n, vector<vector<int>>& edges) {
        size = n;
        mem.resize(n, -1);
        dist.resize(n, INT_MAX);
        dist[n - 1] = 0;
        map.resize(n);
        vector<int> visit(n, 0);
        for(auto edge: edges){
            map[edge[0] - 1][edge[1] - 1] = edge[2];
            map[edge[1] - 1][edge[0] - 1] = edge[2];
        }
        priority_queue<pii, vector<pii>, cmp> pq;
        pq.push({n - 1, 0});
        while(!pq.empty()){
            pii cur = pq.top();
            pq.pop();
            int id = cur.first;
            int dis = cur.second;
            if(visit[id]){
                continue;
            }
            visit[id] = 1;
            for(auto it = map[id].begin(); it != map[id].end(); it++){
                int id2 = it -> first;
                if(visit[id2]){
                    continue;
                }
                dist[id2] = min(dist[id2], it -> second + dis);
                pq.push({id2, dist[id2]});
            }
        }
        int cnt = (int) findpaths(0, n - 1);
        return cnt;
    }

    ll findpaths(int start, int end){
        if(start == end){
            return 1;
        }
        if(mem[start] != -1){
            return mem[start];
        }
        ll cnt = 0;
        for(auto it = map[start].begin(); it != map[start].end(); it++){
            int next = it -> first;
            if(dist[next] < dist[start]){
                cnt += findpaths(next, end);
                cnt = cnt % mod;
            }
        }
        mem[start] = cnt;
        return cnt;
    }
};

int main(){
    Solution5699 s;
    vector<vector<int>> edges = {{1,2,3},{1,3,3},{2,3,1},{1,4,2},{5,2,2},{3,5,1},{5,4,10}};
    cout << s.countRestrictedPaths(5, edges);
}